core/apple/src/SinksApple.kt

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.posix.uint8_tVar
+import kotlin.native.ref.WeakReference
+
+/**
+ * Returns an output stream that writes to this sink. Closing the stream will also close this sink.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Sink.asNSOutputStream(): NSOutputStream = SinkNSOutputStream(this)
+
+@OptIn(UnsafeNumber::class)
+private class SinkNSOutputStream(
+    private val sink: Sink
+) : NSOutputStream(toMemory = Unit), NSStreamDelegateProtocol {
+
+    private val isClosed: () -> Boolean = when (sink) {
+        is RealSink -> sink::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    private var status = NSStreamStatusNotOpen
+    private var error: NSError? = null
+        set(value) {
+            status = NSStreamStatusError
+            field = value
+            postEvent(NSStreamEventErrorOccurred)
+            sink.close()
+        }
+
+    override fun streamStatus() = if (isClosed()) NSStreamStatusClosed else status
+
+    override fun streamError() = error
+
+    override fun open() {
+        if (status == NSStreamStatusNotOpen) {
+            status = NSStreamStatusOpen
+            postEvent(NSStreamEventOpenCompleted)
+            postEvent(NSStreamEventHasSpaceAvailable)
+        }
+    }
+
+    override fun close() {
+        if (status == NSStreamStatusError) return
+        status = NSStreamStatusClosed
+        sink.close()
+        runLoop = null
+        runLoopModes = listOf()
+    }
+
+    @OptIn(DelicateIoApi::class)
+    override fun write(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+        try {
+            if (isClosed()) throw IOException("Underlying sink is closed.")
```

Comment:
I guess we don't have any checked exceptions being thrown now. The goal of the try/catch is to ensure we don't throw any exception during the I/O operation, since the `NSStream` API expects errors to be propagated through the `streamError` property. I narrowed the try/catch to the `write` operation. I'd think the most likely place we could get an exception would probably be with the `hintEmit` call in `writeToInternalBuffer`, which may flush data to the underlying `Sink` and might cause an `IOException`.